sentence_transformers/util.py

```diff
@@ -321,6 +327,308 @@ def semantic_search(
+
+    :return: A tuple containing a list of search results and the time taken for the search. If `output_index` is True,
+        the tuple will also contain the usearch index used for the search.
+    :rtype: Tuple[List[List[Dict[str, Union[int, float]]]], float] or Tuple[List[List[Dict[str, Union[int, float]]]], float, "usearch.index.Index"]
+    :raises ValueError: If both `corpus_embeddings` and `corpus_index` are provided or if neither is provided.
+
+    The list of search results is in the format: [[{"corpus_id": int, "score": float}, ...], ...]
+    The time taken for the search is a float value.
+    """
+    from usearch.index import Index
+    from usearch.compiled import ScalarKind
+
+    if corpus_embeddings is not None and corpus_index is not None:
+        raise ValueError("Only corpus_embeddings or corpus_index should be used, not both.")
+    if corpus_embeddings is None and corpus_index is None:
+        raise ValueError("Either corpus_embeddings or corpus_index should be used.")
+    if corpus_precision not in ["float32", "int8", "binary"]:
+        raise ValueError('corpus_precision must be "float32", "int8", or "binary" for usearch')
+
+    # If corpus_index is not provided, create a new index
+    if corpus_index is None:
+        if corpus_precision == "float32":
+            corpus_index = Index(
+                ndim=corpus_embeddings.shape[1],
+                metric="cos",
+                dtype="f32",
+            )
+        elif corpus_precision == "int8":
+            corpus_index = Index(
+                ndim=corpus_embeddings.shape[1],
+                metric="ip",
+                dtype="i8",
+            )
+        elif corpus_precision == "binary":
+            corpus_index = Index(
+                ndim=corpus_embeddings.shape[1],
+                metric="hamming",
+                dtype="i8",
+            )
+        corpus_index.add(np.arange(len(corpus_embeddings)), corpus_embeddings)
+
+    # If reranking is enabled and the query embeddings are in float32, we need to quantize them
+    # to the same precision as the corpus embeddings. Also update the top_k value to account for the
+    # oversampling factor
+    rerank_embeddings = None
+    k = top_k
+    if query_embeddings.dtype == np.float32:
+        if corpus_index.dtype != ScalarKind.F32 and rerank:
+            rerank_embeddings = query_embeddings
+            k *= oversampling
+
+        query_embeddings = quantize_embeddings(
+            query_embeddings,
+            precision=corpus_precision,
+            ranges=ranges,
+            calibration_embeddings=calibration_embeddings,
+        )
+
+    # Perform the search using the usearch index
+    start_t = time.time()
+    matches = corpus_index.search(query_embeddings, count=k, exact=exact)
+    scores = matches.distances
+    indices = matches.keys
+
+    # If reranking is enabled, we need to rerank the results using the rerank_embeddings
+    if rerank_embeddings is not None:
```

Comment:
There are cases where `rerank_embeddings=None` but `rerank=True`. This might be a big confusing; should we raise some error early on?

Reply:
Resolved via bf90340e5a8e531c9dd1b2652e12f68f3c061817